### PR TITLE
[SPARK-22962][K8S] Fail fast if submission client local files are used

### PR DESCRIPTION
**Pull request by liyinan926**
## What changes were proposed in this pull request?

In the Kubernetes mode, fails fast in the submission process if any submission client local dependencies are used as the use case is not supported yet. 

## How was this patch tested?

Unit tests, integration tests, and manual tests.

@vanzin @foxish

### CODE REVIEW

File: docs/running-on-kubernetes.md [hunk 1]
```diff
@@ -117,7 +117,10 @@ This URI is the location of the example jar that is already in the Docker image.
 If your application's dependencies are all hosted in remote locations like HDFS or HTTP servers, they may be referred to
 by their appropriate remote URIs. Also, application dependencies can be pre-mounted into custom-built Docker images.
 Those dependencies can be added to the classpath by referencing them with `local://` URIs and/or setting the
-`SPARK_EXTRA_CLASSPATH` environment variable in your Dockerfiles.
+`SPARK_EXTRA_CLASSPATH` environment variable in your Dockerfiles. The `local://` scheme is also required when referring to
+dependencies in custom-built Docker images in `spark-submit`. Note that using application dependencies local to the submission
+client is currently not yet supported.
```
Comment (foxish) [inline docs/running-on-kubernetes.md]: "application dependencies from the local file system"?

File: docs/running-on-kubernetes.md [hunk 2]
```diff
@@ -117,7 +117,10 @@ This URI is the location of the example jar that is already in the Docker image.
 If your application's dependencies are all hosted in remote locations like HDFS or HTTP servers, they may be referred to
 by their appropriate remote URIs. Also, application dependencies can be pre-mounted into custom-built Docker images.
 Those dependencies can be added to the classpath by referencing them with `local://` URIs and/or setting the
-`SPARK_EXTRA_CLASSPATH` environment variable in your Dockerfiles.
+`SPARK_EXTRA_CLASSPATH` environment variable in your Dockerfiles. The `local://` scheme is also required when referring to
```
Comment (foxish) [inline docs/running-on-kubernetes.md]: The point above already covers that `local://` is needed with custom-built images.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala [hunk 3]
```diff
@@ -117,6 +117,12 @@ private[spark] class DriverConfigOrchestrator(
       .map(_.split(","))
       .getOrElse(Array.empty[String])
 
+    if (existSubmissionLocalFiles(sparkJars) || existSubmissionLocalFiles(sparkFiles)) {
```
Comment (jiangxb1987) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: We can add a `TODO` here if this is planned to be supported in the future.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala [hunk 4]
```diff
@@ -117,6 +117,12 @@ private[spark] class DriverConfigOrchestrator(
...
+      throw new SparkException("The Kubernetes mode does not yet support application " +
+        "dependencies local to the submission client. It currently only allows application" +
```
Comment (jiangxb1987) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: nit: extra space in the end of line

**Review by jiangxb1987 — APPROVED** [review]
LGTM only some nits.

**Review by foxish — COMMENTED** [review]
+1 on docs change for 2.3 modulo minor comments.
Code changes ok if we also re-validate with manual tests against http (we have a caveat with our integration tests not testing this correctly yet), gcs and HDFS. Also ok with dropping the code change for 2.3, since it's a usability and not a functionality improvement.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala (hunk 3 above)
Comment (liyinan926) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: Done, and also created https://issues.apache.org/jira/browse/SPARK-23153.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala (hunk 4 above)
Comment (liyinan926) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: Done.

File: docs/running-on-kubernetes.md (hunk 2 above)
Comment (liyinan926) [reply docs/running-on-kubernetes.md]: Yes, but that's about adding to the classpath. I wanted to make it more specific and clearer.

File: docs/running-on-kubernetes.md (hunk 1 above)
Comment (liyinan926) [reply docs/running-on-kubernetes.md]: Done.

**Comment by liyinan926** [comment]
Regarding manual tests, our integration tests cover `local://` and `http://` (at least for files). I will do some manual tests on `http://` for jars and `gs://`.

**Comment by liyinan926** [comment]
Manual tests to verify that using `http://`/`https://`, `gcs://`, and `local://` dependencies works fine. Also verified that using `file://` dependencies results in submission failure due to the exception.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala [hunk 5]
```diff
@@ -117,6 +117,13 @@ private[spark] class DriverConfigOrchestrator(
       .map(_.split(","))
       .getOrElse(Array.empty[String])
 
+    // TODO(SPARK-23153): remote once submission client local dependencies are supported.
```
Comment (foxish) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: s/remote/remove/

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala [hunk 6]
```diff
@@ -117,6 +117,13 @@ private[spark] class DriverConfigOrchestrator(
...
+    if (existSubmissionLocalFiles(sparkJars) || existSubmissionLocalFiles(sparkFiles)) {
+      throw new SparkException("The Kubernetes mode does not yet support application " +
```
Comment (foxish) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: I'd shorten this to just "Kubernetes mode does not support referencing application dependencies in the local file system".

**Review by foxish — COMMENTED** [review]
Question - is referencing a local file as a main app resource (`/home/../spark-examples.jar` for example) need to be dealt with separately?

**Comment by liyinan926** [comment]
Actually main application jar gets added to `spark.jars` automatically by the submission process so are treated the same as other jars on `spark.jars`.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala (hunk 5 above)
Comment (liyinan926) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: Done.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala (hunk 6 above)
Comment (liyinan926) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigOrchestrator.scala]: Done.

**Comment by foxish** [comment]
Suspected that it would use the same code path - might be worth a manual test of that as well. 
LGTM from my end.

**Comment by liyinan926** [comment]
The manual tests I did actually use a main app jar located on gcs and http.

**Comment by jiangxb1987** [comment]
LGTM

**Comment by vanzin** [comment]
Merging to master / 2.3.
